core/drivers/imx_snvs.c

```diff
@@ -2,37 +2,191 @@
+	uint32_t lp_lr = 0;
+	vaddr_t base = core_mmu_get_va(SNVS_BASE, MEM_AREA_IO_SEC, SNVS_SIZE);
+
+	hp_lr = io_read32(base + SNVS_HPLR);
+	lp_lr = io_read32(base + SNVS_LPLR);
+
+	if ((hp_lr & SNVS_HPLR_MKS_SL) || (lp_lr & SNVS_LPLR_MKS_HL))
+		return true;
+
+	return false;
+}
+
+/*
+ * Set the Master key to use OTPMK and lock it. Return true if the operation is
+ * successful. Return false if the master key is already locked.
+ */
+static void set_mks_otpmk(void)
+{
+	vaddr_t base = core_mmu_get_va(SNVS_BASE, MEM_AREA_IO_SEC, SNVS_SIZE);
+
+	io_setbits32(base + SNVS_HPCOMR, SNVS_HPCOMR_MKS_EN);
+	io_clrbits32(base + SNVS_LPMKCR, SNVS_LPMKCR_MKCR_MKS_SEL);
+	io_clrbits32(base + SNVS_HPLR, SNVS_HPLR_MKS_SL);
+	io_setbits32(base + SNVS_LPLR, SNVS_LPLR_MKS_HL);
+}
+
+/*
+ * Return true if OTPMK is valid, false otherwise.
+ */
+static bool is_otpmk_valid(void)
+{
+	uint32_t hp_status = 0;
+	vaddr_t base = core_mmu_get_va(SNVS_BASE, MEM_AREA_IO_SEC, SNVS_SIZE);
+
+	hp_status = io_read32(base + SNVS_HPSR);
+
+	if (hp_status & (SNVS_HPSR_OTPMK_ZERO | SNVS_HPSR_OTPMK_SYND))
+		return false;
+
+	return true;
+}
+
+#ifdef CFG_MX8M
+#define SNVS_HPSR_SYS_SECURITY_CFG GENMASK_32(15, 12)
+
+static enum snvs_security_cfg snvs_get_security_cfg(void)
 {
-	vaddr_t snvs = core_mmu_get_va(SNVS_BASE, MEM_AREA_IO_SEC,
-				       SNVS_HPSR + sizeof(uint32_t));
 	uint32_t val = 0;
+	vaddr_t base = core_mmu_get_va(SNVS_BASE, MEM_AREA_IO_SEC,
+				       SNVS_SIZE);
 
-	val = io_read32(snvs + SNVS_HPSR);
-	DMSG("HPSR: 0x%"PRIx32, val);
-	if (val & SNVS_HPSR_SYS_SECURITY_BAD)
-		return SNVS_SECURITY_CFG_FIELD_RETURN;
-	else if (val & SNVS_HPSR_SYS_SECURITY_CLOSED)
-		return SNVS_SECURITY_CFG_CLOSED;
-	else if (val & SNVS_HPSR_SYS_SECURITY_OPEN)
+	val = io_read32(base + SNVS_HPSR) &
+	      SNVS_HPSR_SYS_SECURITY_CFG >> SNVS_HPSR_SYS_SECURITY_CFG_OFFSET;
+
+	switch (val) {
+	case 0b0000:
+	case 0b1000:
+		return SNVS_SECURITY_CFG_FAB;
+	case 0b0001:
+	case 0b0010:
+	case 0b0011:
 		return SNVS_SECURITY_CFG_OPEN;
-	else if (val > 4 && val < 8)
+	case 0b1010:
+	case 0b1001:
+	case 0b1011:
+		return SNVS_SECURITY_CFG_CLOSED;
+	default:
+		return SNVS_SECURITY_CFG_FIELD_RETURN;
+	}
+}
+#else
+#define SNVS_HPSR_SYS_SECURITY_CFG GENMASK_32(14, 12)
+
+static enum snvs_security_cfg snvs_get_security_cfg(void)
+{
+	uint32_t val = 0;
+	vaddr_t base = core_mmu_get_va(SNVS_BASE, MEM_AREA_IO_SEC,
+				       SNVS_SIZE);
+
+	val = io_read32(base + SNVS_HPSR) &
+	      SNVS_HPSR_SYS_SECURITY_CFG >> SNVS_HPSR_SYS_SECURITY_CFG_OFFSET;
+
+	switch (val) {
+	case 0b000:
+		return SNVS_SECURITY_CFG_FAB;
+	case 0b001:
 		return SNVS_SECURITY_CFG_OPEN;
+	case 0b011:
+		return SNVS_SECURITY_CFG_CLOSED;
+	default:
+		return SNVS_SECURITY_CFG_FIELD_RETURN;
+	}
+}
+#endif
 
-	return SNVS_SECURITY_CFG_FAB;
+static bool snvs_is_device_closed(void)
+{
+	return (snvs_get_security_cfg() == SNVS_SECURITY_CFG_CLOSED);
 }
 
-enum snvs_ssm_mode snvs_get_ssm_mode(void)
+#ifdef CFG_RPMB_FS
```

Comment:
Shouldn't this compile without the `#ifdef`?

Reply:
`snvs_get_ssm_mode()` is only used in `plat_rpmb_key_is_ready()` and the latter is implemented only if `CFG_RPMB_FS=y`